### PR TITLE
Allow log_tags to take a hash for a tagged logger

### PR DESCRIPTION
**Pull request by allcentury**
### Summary

Tagged loggers like `ActiveSupport::TaggedLogging` make logging both named
and unnamed tags visible for better application debugging. Prior to this
change, log_tags only supported an array of tags such as:

```ruby
config.log_tags = [
  :remote_ip,
  ->(request) { some_computed_value }
]
```

However as this list gets long it has two issues:

1.  Parsing JSON logs requires positional references (ie remote_ip will
always be at some index)
2.  Some tags look similar and are hard to discern in a log message.
For example a datadog trace id and epoch time look indistinguishable to a
human reading those logs

This change allows users to instead pass in a hash like object to
`#log_tags`:

```ruby
config.log_tags = {
  remote_ip: :remote_ip,
  trace_id: ->(request) { some_computed_value }
}
```

### Other Information

- I still need to update the changelog if folks like this.

### CODE REVIEW

File: railties/lib/rails/rack/logger.rb [hunk 1]
```diff
@@ -23,7 +23,8 @@ def call(env)
         request = ActionDispatch::Request.new(env)
 
         if logger.respond_to?(:tagged)
-          logger.tagged(compute_tags(request)) { call_app(request, env) }
+          tags = compute_tags(request)
+          logger.tagged(tags) { call_app(request, env) }
```
Comment (eugeneius) [inline railties/lib/rails/rack/logger.rb]: This change seems unnecessary.

**Comment by eugeneius** [comment]
What does `ActiveSupport::TaggedLogger` write to the log when given tags as a hash?

**Comment by allcentury** [comment]
@eugeneius with a simple formatter by default it prints:

```
[{:ip=>"127.0.0.1", :source=>"my-app", :trace_id=>0, :span_id=>0}] Started GET "/search" for 127.0.0.1 at 2019-12-17 19:07:00 +0000
```

With a JSON formatter, it prints as such:

```json
{
  "level": "INFO",
  "time": "2019-12-17T21:25:41.997869+00:00",
  "msg": "Completed 200 OK in 47ms (Views: 2.2ms | ActiveRecord: 4.4ms)",
  "progname": "app-test[2984]",
  "tags": [
    {
      "external_request_id": "no-request-id                       ",
      "local_request_id": "71ee2ffe-dd9e-435f-97c5-5d1d8f822748"
    },
    {
      "ip": "127.0.0.1",
      "source": "my-app",
      "trace_id": 0,
      "span_id": 0
    }
  ]
}
```

**Comment by eugeneius** [comment]
The fact that the default formatter will store a hash as a single tag and dump it into the log line without crashing seems more like a happy accident than an intentionally designed behaviour.

I'm interested to hear @georgeclaghorn's opinion, but I think if this feature is going to be merged it has to work well with the default formatter out of the box, with output closer to something like this:

```
[ip:127.0.0.1] [source:my-app] [trace_id:0] [span_id:0] Started GET "/search" for 127.0.0.1 at 2019-12-17 19:07:00 +0000
```

**Comment by allcentury** [comment]
@eugeneius - I think that's a good proposal, I'd be happy to implement that.  Let me know your thoughts @georgeclaghorn

**Comment by allcentury** [comment]
Friendly bump @georgeclaghorn

**Comment by allcentury** [comment]
@georgeclaghorn - friendly bump again

**Comment by allcentury** [comment]
friendly bump @georgeclaghorn

**Comment by allcentury** [comment]
👋

**Comment by md5** [comment]
I'm wondering if it would make sense to update `TaggedLogging#tagged` to accept kwargs in addition to the existing splat args. This could make it less likely to break existing extensions and monkey patching of `TaggedLogging`.

I'm looking at being able to log the same sort of things that `log_tags` currently configures to be pulled from `request` in `Rails::Rack::Logger#compute_tags` as JSON key-value pairs, but it looks like in order to do so I'd have to entirely reproduce the functionality in both `TaggedLogging` and `Rails::Rack::Logger` to make it compatible with a key-value format. For messages that are intercepted and replaced with `lograge`, it's possible to use `custom_options` and `custom_payload` to capture the same information from `request`, but my goal is to have all log messages to `Rails.logger` be tagged with k/v pairs as currently happens with just the values that are prepended by `TaggedLogging`.

**Comment by md5** [comment]
After working with `TaggedLogging` a bit more, I was able to get all the tags logged in a JSON `tags` array while still using `TaggedLogging` directly. The only thing I had to monkey patch was to avoid the override of `#call` in `TaggedLogging::Formatter`, due to the inline interpolation of `tags_text` as `"#{tags_text}#{msg}"`. If this interpolation were broken out into a method on `Formatter` instead of happening inline in the definition of `call`, it would be easier to override it to be a no-op. This would allow a custom `Formatter` to still benefit from `current_tags` and the behavior of `Rails::Rack::Logger` calling `tagged`. Here's what I had to do to work around the inline interpolation in `#call`:

```ruby
class JsonFormatter < ::Logger::Formatter
  def call(severity, timestamp, progname, msg)
    msg = { message: msg } if msg.is_a?(String)

    json = JSON.generate(
      severity: severity,
      timestamp: format_datetime(timestamp),
      progname: progname,
      msg: msg,
      tags: current_tags
    )

    "#{json}\n"
  end

  alias call_without_tagging call
end

# Suppresses prefixing of tags_text onto log messages since we only want tags
# in the "tags" array of the JSON log object
module SuppressTagsText
  def call(severity, timestamp, progname, msg)
    call_without_tagging(severity, timestamp, progname, msg)
  end
end

# ...

logger = ActiveSupport::Logger.new(io)

# Configure JSON log formatter
logger.formatter = JsonFormatter.new

# Enable tagged logging to allow Rails to pass values from
# Rails::Rack::Logger
app.config.logger = ActiveSupport::TaggedLogging.new(logger)

# Disable tags_text prefixing done by TaggedLogging
app.config.logger.formatter.extend SuppressTagsText
```

If instead of doing `super(severity, timestamp, progname, "#{tags_text}#{msg}")`, `TaggedLogging` instead did `super(severity, timestamp, progname, tag_message(msg))`, I could override `tag_message` to simply return `msg` and avoid the `alias` hack.

**Pull request closed without merging**